routing_plugin.go

```diff
@@ -104,3 +115,322 @@ func ReleaseRoutingPlugin(ctx context.Context) {
 		routingPluginInstance = nil
 	}
 }
+
+// lowToHighRoutingPlugin is a RoutingPlugin that implements "low to high"
+// routing. This means that when we're attempting to pay to a target we'll
+// gradually (with a linear step function) discard inbound peers to that target
+// given routing timeouts. The lowToHighRoutingPlugin itself is responsible for
+// manipulating LND's Mission Control to make such routing attempts possible.
+type lowToHighRoutingPlugin struct {
+	lnd     lndclient.LndServices
+	target  route.Vertex
+	amount  btcutil.Amount
+	mcState []lndclient.MissionControlEntry
+
+	// nodesByMaxFee holds nodes sorted by maximum fees that would be paid
+	// to the target node for the target amount.
+	nodesByMaxFee []nodeFeeInfo
+}
+
+type nodeFeeInfo struct {
+	node     route.Vertex
+	capacity btcutil.Amount
+	fee      int64
+}
+
+// Enforce that lowToHighRoutingPlugin implements the RoutingPlugin interface.
+var _ RoutingPlugin = (*lowToHighRoutingPlugin)(nil)
+
+// Init will initialize the "low to high" routing plugin. It'll save the MC
+// state and also preinit the internal state of the routing plugin. When the
+// instance is released, the saved MC state can be restored.
+func (r *lowToHighRoutingPlugin) Init(ctx context.Context,
+	target route.Vertex, amt btcutil.Amount) error {
+
+	// Save the original target as the current "exit" node: where
+	// our payments should flow forward.
+	exit := target
+
+	// Walk until the first fork (if there's any). This first
+	// fork will be where we're going to try to manipulate success
+	// probabilities to increasingly prefer more expensive edges.
+
+	// We track all visited peers, so we won't end up walking
+	// back and forth on graphs that have no forks.
+	visited := make(map[route.Vertex]struct{})
+	for {
+		targetNodeInfo, err := r.lnd.Client.GetNodeInfo(
+			ctx, target, true,
```

Comment:
Is target the primary? Are we going to be able to find it in the graph if it's all private channels?

Reply:
Ah nvm, we have hop hints so it should be there, obviously 🤦‍♀️

Reply:
Realized that indeed the original solution wouldn't work with private channels. With the latest version edges from the hop hints in the invoice are used too.